File: anydialog/src/main/java/per/goweii/anydialog/AnyDialog.java
package per.goweii.anydialog;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.FloatRange;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageView;

import per.goweii.burred.Blurred;

/**
 * @author CuiZhen
 * @date 2018/4/30
 * QQ: 302833254
 * E-mail: goweii@163.com
 * GitHub: https://github.com/goweii
 */
public class AnyDialog extends Dialog {

    private static final long ANIM_DURATION_DEFAULT = 300L;

    private final ViewHolder mViewHolder;
    private final Context mContext;

    private float mBackgroundDimAmount = -1;
    private int mBackgroundColorInt = -1;
    private float mBackgroundBlurPercent = 0;
    private float mBackgroundBlurRadius = 0;
    private float mBackgroundBlurScale = 2;

    private IAnim mContentAnim = null;
    private AnimatorSet mInAnim = null;
    private AnimatorSet mOutAnim = null;

    private int mGravity = -1;
    private boolean mCancelableOnTouchOutside = true;

    private boolean mInsideStatusBar = false;
    private boolean mInsideNavigationBar = false;

    private int[] mInsideParamsOff = new int[]{0, 0, 0, 0};

    private IDataBinder mDataBinder = null;

    private OnDialogCreatedListener mOnDialogCreatedListener = null;
    private OnDialogEnterListener mOnDialogEnterListener = null;
    private OnDialogShownListener mOnDialogShownListener = null;
    private OnDialogExitListener mOnDialogExitListener = null;
    private OnDialogDismissedListener mOnDialogDismissedListener = null;

    public static AnyDialog with(@NonNull Context context) {
        return new AnyDialog(context);
    }

    private AnyDialog(@NonNull Context context) {
        super(context);
        this.mContext = context;
        this.mViewHolder = new ViewHolder(this);
        Blurred.init(mContext);
    }

    @Override
    public void show() {
        super.show();
    }

    @Override
    public void dismiss() {
        onDialogExit();
    }

    public AnyDialog contentAnim(IAnim contentAnim) {
        this.mContentAnim = contentAnim;
        return this;
    }

    public AnyDialog contentView(@NonNull View contentView) {
        mViewHolder.setContent(contentView);
        return this;
    }

    public AnyDialog contentView(@LayoutRes int contentViewId) {
        return contentView(LayoutInflater.from(mContext).inflate(contentViewId, mViewHolder.getContainer(), false));
    }

    /**
     * 背景变暗程度
     *
     * @param dimAmount 变暗程度
     * @return AnyDialog
     */
    public AnyDialog backgroundDimAmount(@FloatRange(from = 0, to = 1) float dimAmount) {
        this.mBackgroundDimAmount = dimAmount;
        return this;
    }

    /**
     * 背景色
     *
     * @param colorInt 背景色
     * @return AnyDialog
     */
    public AnyDialog backgroundColorInt(@ColorInt int colorInt) {
        this.mBackgroundColorInt = colorInt;
        return this;
    }

    /**
     * 背景色
     *
     * @param colorRes 背景色
     * @return AnyDialog
     */
    public AnyDialog backgroundColorRes(@ColorRes int colorRes) {
        this.mBackgroundColorInt = ContextCompat.getColor(mContext, colorRes);
        return this;
    }

    public AnyDialog backgroundBlurPercent(float backgroundBlurPercent) {
        mBackgroundBlurPercent = backgroundBlurPercent;
        return this;
    }

    public AnyDialog backgroundBlurRadius(float backgroundBlurRadius) {
        mBackgroundBlurRadius = backgroundBlurRadius;
        return this;
    }

    public AnyDialog backgroundBlurScale(float backgroundBlurScale) {
        mBackgroundBlurScale = backgroundBlurScale;
        return this;
    }

    public AnyDialog cancelableOnClickKeyBack(boolean cancelable) {
        setCancelable(cancelable);
        return this;
    }

    public AnyDialog cancelableOnTouchOutside(boolean cancelable) {
        setCanceledOnTouchOutside(cancelable);
        this.mCancelableOnTouchOutside = cancelable;
        return this;
    }

    public AnyDialog insideStatusBar(boolean insideStatusBar) {
        this.mInsideStatusBar = insideStatusBar;
        return this;
    }

    public AnyDialog insideNavigationBar(boolean insideNavigationBar) {
        this.mInsideNavigationBar = insideNavigationBar;
        return this;
    }

    /**
     * {@link android.view.Gravity}
     *
     * @param gravity Gravity
     * @return AnyDialog
     */
    public AnyDialog gravity(int gravity) {
        this.mGravity = gravity;
        return this;
    }

    public AnyDialog bindData(IDataBinder dataBinder) {
        this.mDataBinder = dataBinder;
        return this;
    }

    public AnyDialog onClick(@IdRes int viewId, OnDialogClickListener listener) {
        mViewHolder.addOnClickListener(viewId, listener);
        return this;
    }

    public AnyDialog onClick(OnDialogClickListener listener, @IdRes int... viewIds) {
        if (viewIds == null || viewIds.length == 0) {
            return this;
        }
        for (int viewId : viewIds) {
            mViewHolder.addOnClickListener(viewId, listener);
        }
        return this;
    }

    public AnyDialog onClickToDismiss(@IdRes int... viewIds) {
        return onClick(new OnDialogClickListener() {
            @Override
            public void onClick(AnyDialog anyDialog, View view) {
                dismiss();
            }
        }, viewIds);
    }

    public AnyDialog onDialogCreatedListener(OnDialogCreatedListener onDialogCreatedListener) {
        mOnDialogCreatedListener = onDialogCreatedListener;
        return this;
    }

    public AnyDialog setOnDialogEnterListener(OnDialogEnterListener onDialogEnterListener) {
        mOnDialogEnterListener = onDialogEnterListener;
        return this;
    }

    public AnyDialog setOnDialogShownListener(OnDialogShownListener onDialogShownListener) {
        mOnDialogShownListener = onDialogShownListener;
        return this;
    }

    public AnyDialog setOnDialogExitListener(OnDialogExitListener onDialogExitListener) {
        mOnDialogExitListener = onDialogExitListener;
        return this;
    }

    public AnyDialog setOnDialogDismissedListener(OnDialogDismissedListener onDialogDismissedListener) {
        mOnDialogDismissedListener = onDialogDismissedListener;
        return this;
    }

    public <V extends View> V getView(@IdRes int viewId) {
        return mViewHolder.getView(viewId);
    }

    public ViewHolder getViewHolder() {
        return mViewHolder;
    }

    public FrameLayout getContainer() {
        return mViewHolder.getContainer();
    }

    public ImageView getBackground() {
        return mViewHolder.getBackground();
    }

    public View getContentView() {
        return mViewHolder.getContent();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(mViewHolder.getContainer());
        initWindow();

        initContainer();
        calculateInsideParamsOff();
        initBackground();
        initContent();
        mViewHolder.bindListener();

        onDialogCreated();
    }

    private void initWindow() {
        Window window = getWindow();
        if (window == null) {
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.TRANSPARENT);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        }

        window.setWindowAnimations(0);
        window.clearFlags(WindowManager.LayoutParams.DIM_AMOUNT_CHANGED);
        WindowManager.LayoutParams layoutParams = window.getAttributes();
        layoutParams.width = WindowManager.LayoutParams.MATCH_PARENT;
        layoutParams.height = WindowManager.LayoutParams.MATCH_PARENT;
        layoutParams.dimAmount = 0;
        window.setAttributes(layoutParams);

        View decorView = window.getDecorView();
        decorView.setPadding(0, 0, 0, 0);
        decorView.setBackgroundColor(Color.TRANSPARENT);
    }

    protected void onDialogCreated() {
        if (mOnDialogCreatedListener != null) {
            mOnDialogCreatedListener.onCreated(AnyDialog.this);
        }
        if (mDataBinder != null) {
            mDataBinder.bind(AnyDialog.this);
        }
        mViewHolder.getContainer().post(new Runnable() {
            @Override
            public void run() {
                onDialogEnter();
            }
        });
    }

    protected void onDialogEnter() {
        if (mOnDialogEnterListener != null) {
            mOnDialogEnterListener.onEnter(AnyDialog.this);
        }
        startInAnim();
    }

    protected void onDialogShown() {
        if (mOnDialogShownListener != null) {
            mOnDialogShownListener.onShown(AnyDialog.this);
        }
    }

    protected void onDialogExit() {
        if (mOnDialogExitListener != null) {
            mOnDialogExitListener.onExit(AnyDialog.this);
        }
        startOutAnim();
    }

    protected void onDialogDismissed() {
        AnyDialog.super.dismiss();
        if (mOnDialogDismissedListener != null) {
            mOnDialogDismissedListener.onDismissed(AnyDialog.this);
        }
    }

    private void initContainer() {
        if (mCancelableOnTouchOutside) {
            mViewHolder.getContainer().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    dismiss();
                }
            });
        }
    }

    private void initBackground() {
        if (mBackgroundBlurPercent > 0 || mBackgroundBlurRadius > 0) {
            Activity activity = Utils.getActivity(mContext);
            if (activity != null) {
                FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) mViewHolder.getBackground().getLayoutParams();
                params.leftMargin += mInsideParamsOff[0];
                params.rightMargin += mInsideParamsOff[2];
                params.bottomMargin += mInsideParamsOff[3];
                final View activityDecor = activity.getWindow().getDecorView();
                Bitmap decorBitmap = Utils.snapshot(activityDecor);
                int x = mInsideParamsOff[0];
                int y = 0;
                int w = decorBitmap.getWidth() - x - mInsideParamsOff[2];
                int h = decorBitmap.getHeight() - y - mInsideParamsOff[3];
                Bitmap cutBitmap = Bitmap.createBitmap(decorBitmap, x, y, w, h);
                decorBitmap.recycle();
                Blurred blurred = Blurred.with(cutBitmap)
                        .recycleOriginal(true)
                        .keepSize(false)
                        .scale(mBackgroundBlurScale);
                if (mBackgroundBlurPercent > 0) {
                    blurred.percent(mBackgroundBlurPercent);
                } else if (mBackgroundBlurRadius > 0) {
                    blurred.radius(mBackgroundBlurRadius);
                }
                Bitmap blurBitmap = blurred.blur();
                if (blurBitmap != null) {
                    BitmapDrawable blurDrawable = new BitmapDrawable(mContext.getResources(), blurBitmap);
                    if (mBackgroundColorInt == -1) {
                        mBackgroundColorInt = Color.TRANSPARENT;
                    }
                    if (mBackgroundDimAmount == -1) {
                        mBackgroundDimAmount = 0;
                    }
                    mViewHolder.getBackground().setImageDrawable(blurDrawable);
                    mViewHolder.getBackground().setColorFilter(Utils.alphaColor(mBackgroundColorInt, mBackgroundDimAmount));
                }
            }
        } else {
            if (mBackgroundColorInt == -1) {
                mBackgroundColorInt = Color.BLACK;
                if (mBackgroundDimAmount == -1) {
                    mBackgroundDimAmount = 0.382F;
                }
            } else {
                if (mBackgroundDimAmount == -1) {
                    mBackgroundDimAmount = 0;
                }
            }
            ColorDrawable colorDrawable = new ColorDrawable(mBackgroundColorInt);
            colorDrawable.setAlpha((int) (mBackgroundDimAmount * 255));
            mViewHolder.getBackground().setImageDrawable(colorDrawable);
        }
    }

    private void initContent() {
        if (mViewHolder.getContent() == null) {
            return;
        }
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) mViewHolder.getContent().getLayoutParams();
        if (mGravity != -1) {
            params.gravity = mGravity;
        } else if (params.gravity == FrameLayout.LayoutParams.UNSPECIFIED_GRAVITY) {
            params.gravity = Gravity.CENTER;
        }
        if (params.width == FrameLayout.LayoutParams.MATCH_PARENT) {
            params.gravity |= Gravity.LEFT;
        }
        if (params.height == FrameLayout.LayoutParams.MATCH_PARENT) {
            params.gravity |= Gravity.TOP;
        }
        if (mInsideStatusBar) {
            params.topMargin += mInsideParamsOff[1];
        }
        if (mInsideNavigationBar) {
            params.leftMargin += mInsideParamsOff[0];
            params.rightMargin += mInsideParamsOff[2];
            params.bottomMargin += mInsideParamsOff[3];
        }
        mViewHolder.getContent().setFocusable(true);
        mViewHolder.getContent().setClickable(true);
        mViewHolder.getContainer().addView(mViewHolder.getContent());
    }

    private void calculateInsideParamsOff() {
        mInsideParamsOff[1] = Utils.getStatusBarHeight(mContext);
        Activity activity = Utils.getActivity(mContext);
        if (activity != null) {
            final View activityDecor = activity.getWindow().getDecorView();
            final View activityContent = activityDecor.findViewById(android.R.id.content);
            int[] activityDecorLocation = new int[2];
            activityDecor.getLocationOnScreen(activityDecorLocation);
            final int adLeft = activityDecorLocation[0];
            final int adTop = activityDecorLocation[1];
            final int adRight = adLeft + activityDecor.getWidth();
            final int adBottom = adTop + activityDecor.getHeight();
            int[] activityContentLocation = new int[2];
            activityContent.getLocationOnScreen(activityContentLocation);
            final int acLeft = activityContentLocation[0];
            final int acTop = activityContentLocation[1];
            final int acRight = acLeft + activityContent.getWidth();
            final int acBottom = acTop + activityContent.getHeight();
            mInsideParamsOff[0] = acLeft - adLeft;
            mInsideParamsOff[2] = -(acRight - adRight);
            mInsideParamsOff[3] = -(acBottom - adBottom);
        }
    }

    private void startInAnim() {
        if (mInAnim != null && mInAnim.isRunning()) {
            return;
        }
        mInAnim = createInAnim();
        mInAnim.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                mInAnim = null;
                onDialogShown();
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                mInAnim = null;
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        mInAnim.start();
    }

    private AnimatorSet createInAnim() {
        Animator contentInAnimator = null;
        if (mContentAnim != null) {
            contentInAnimator = mContentAnim.inAnim(mViewHolder.getContent());
        }
        if (contentInAnimator == null) {
            contentInAnimator = AnimHelper.createZoomInAnim(mViewHolder.getContent());
        }
        if (contentInAnimator.getDuration() < 0) {
            contentInAnimator.setDuration(ANIM_DURATION_DEFAULT);
        }
        Animator backgroundInAnim = AnimHelper.createAlphaInAnim(mViewHolder.getBackground());
        backgroundInAnim.setDuration(contentInAnimator.getDuration());
        AnimatorSet inAnim = new AnimatorSet();
        inAnim.playTogether(contentInAnimator, backgroundInAnim);
        return inAnim;
    }

    private void startOutAnim() {
        if (mOutAnim != null && mOutAnim.isRunning()) {
            return;
        }
        mOutAnim = createOutAnim();
        mOutAnim.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                mOutAnim = null;
                onDialogDismissed();
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                mOutAnim = null;
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        mOutAnim.start();
    }

    private AnimatorSet createOutAnim() {
        Animator contentOutAnimator = null;
        if (mContentAnim != null) {
            contentOutAnimator = mContentAnim.outAnim(mViewHolder.getContent());
        }
        if (contentOutAnimator == null) {
            contentOutAnimator = AnimHelper.createZoomOutAnim(mViewHolder.getContent());
        }
        if (contentOutAnimator.getDuration() < 0) {
            contentOutAnimator.setDuration(ANIM_DURATION_DEFAULT);
        }
        Animator backgroundOutAnim = AnimHelper.createAlphaOutAnim(mViewHolder.getBackground());
        backgroundOutAnim.setDuration(contentOutAnimator.getDuration());
        AnimatorSet outAnim = new AnimatorSet();
        outAnim.playTogether(contentOutAnimator, backgroundOutAnim);
        return outAnim;
    }
}

File: anydialog/src/main/java/per/goweii/anydialog/AnyDialog.kt
package per.goweii.anydialog

import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.Point
import android.os.Build
import android.os.Bundle
import android.view.*
import android.widget.FrameLayout
import androidx.annotation.*

open class AnyDialog(
        context: Context,
        private val initializer: (AnyDialog.() -> Unit)? = null
) : Dialog(context, R.style.Dialog) {

    private val viewHolder = ViewHolder()

    lateinit var container: ViewGroup

    @LayoutRes
    open var contentRes: Int = 0
    open lateinit var content: View

    @StyleRes
    open var animation: Int = R.style.DialogAnimDef

    open var dimBehind: Boolean = true
    @FloatRange(from = 0.0, to = 1.0)
    open var dimAmount: Float? = null

    open var gravity: Int? = null

    open var width: Int? = null
    open var height: Int? = null
    open var fullscreen: Boolean = false

    open var fitSystemWindow: Boolean = false
    open var fitSystemLeft: Boolean = false
    open var fitSystemTop: Boolean = false
    open var fitSystemRight: Boolean = false
    open var fitSystemBottom: Boolean = false

    open var style: Style? = null
        set(value) {
            field = value?.also {
                animation = it.animRes
                gravity = it.gravity
                width = it.width
                height = it.height
                dragDirection = it.dragDirection
            }
        }

    open var dragDirection: DragDirection? = null

    var dataBind: (AnyDialog.() -> Unit)? = null

    @CallSuper
    override fun onCreate(savedInstanceState: Bundle?) {
        initializer?.invoke(this)
        super.onCreate(savedInstanceState)
        container = onCreateContainer()
        content = onCreateView(container, savedInstanceState)
        container.layoutParams = content.layoutParams
        content.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        container.addView(content)
        viewHolder.attach(container)
        onViewCreated(container, savedInstanceState)
        setContentView(container)
        onViewAttached(container, savedInstanceState)
    }

    @CallSuper
    protected open fun onCreateContainer(): ViewGroup {
        val parent = window!!.decorView.findViewById<ViewGroup>(Window.ID_ANDROID_CONTENT)
        return dragDirection?.let {
            DragLayout3(parent.context).apply {
                dragStyle(when (it) {
                    DragDirection.LEFT -> DragLayout3.DragStyle.Left
                    DragDirection.TOP -> DragLayout3.DragStyle.Top
                    DragDirection.RIGHT -> DragLayout3.DragStyle.Right
                    DragDirection.BOTTOM -> DragLayout3.DragStyle.Bottom
                })
                onDragging { f ->
                    dimAmount?.let {
                        window?.setDimAmount(it * (1F - f))
                    }
                }
                onDragEnd { dismiss() }
            }
        } ?: run {
            FrameLayout(parent.context)
        }
    }

    @CallSuper
    protected open fun onCreateView(container: ViewGroup, savedInstanceState: Bundle?): View {
        return if (this::content.isInitialized) {
            this.content
        } else {
            layoutInflater.inflate(contentRes, container, false)
        }
    }

    @CallSuper
    protected open fun onViewCreated(content: View, savedInstanceState: Bundle?) {
        if (fitSystemWindow || fitSystemLeft || fitSystemTop || fitSystemRight || fitSystemBottom) {
            content.fitsSystemWindows = true
            if (content is ViewGroup) {
                content.clipToPadding = false
                if (!fitSystemWindow && (!fitSystemLeft || !fitSystemTop || !fitSystemRight || !fitSystemBottom)) {
                    content.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
                        override fun onPreDraw(): Boolean {
                            if (content.viewTreeObserver.isAlive) {
                                content.viewTreeObserver.removeOnPreDrawListener(this)
                            }
                            val l = if (fitSystemLeft) content.paddingLeft else 0
                            val t = if (fitSystemTop) content.paddingTop else 0
                            val r = if (fitSystemRight) content.paddingRight else 0
                            val b = if (fitSystemBottom) content.paddingBottom else 0
                            content.setPadding(l, t, r, b)
                            return true
                        }
                    })
                }
            }
        }
        val params = content.layoutParams as FrameLayout.LayoutParams
        if (gravity == null && params.gravity != FrameLayout.LayoutParams.UNSPECIFIED_GRAVITY) {
            gravity = params.gravity
        }
        if (width == null || height == null) {
            Point().apply {
                window!!.windowManager.defaultDisplay.getSize(this)
            }.also { size ->
                if (width == null) {
                    width = if (params.width == ViewGroup.LayoutParams.MATCH_PARENT) {
                        if (params.leftMargin == 0 && params.rightMargin == 0) {
                            params.width
                        } else {
                            size.x - params.leftMargin - params.rightMargin
                        }
                    } else {
                        params.width
                    }
                }
                if (height == null) {
                    height = if (params.height == ViewGroup.LayoutParams.MATCH_PARENT) {
                        if (params.topMargin == 0 && params.bottomMargin == 0) {
                            params.height
                        } else {
                            size.y - params.topMargin - params.bottomMargin
                        }
                    } else {
                        params.height
                    }
                }
            }
        }
        dataBind?.invoke(this)
        viewHolder.bindClickListener(this)
    }

    @CallSuper
    protected open fun onViewAttached(content: View, savedInstanceState: Bundle?) {
        val w = window!!
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            w.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            w.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            w.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            w.statusBarColor = Color.TRANSPARENT
        }
        w.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN)
        val decor: View = w.decorView
        decor.setPadding(0, 0, 0, 0)
        decor.setBackgroundColor(Color.TRANSPARENT)
        w.setWindowAnimations(animation)
        gravity?.let { w.attributes.gravity = it }
        width?.let { w.attributes.width = it }
        height?.let { w.attributes.height = it }
        if (fullscreen) {
            w.attributes.width = WindowManager.LayoutParams.MATCH_PARENT
            w.attributes.height = WindowManager.LayoutParams.MATCH_PARENT
        }
    }

    override fun onStart() {
        resetDim()
        super.onStart()
    }

    private fun resetDim() {
        val w = window!!
        if (dimBehind) {
            w.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        } else {
            w.clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        }
        dimAmount?.let {
            w.setDimAmount(it)
        } ?: run {
            dimAmount = w.attributes.dimAmount
        }
    }

    /**
     * 是否可背景变暗，所有dialog公用同一个背景遮罩
     * 为false时dimAmount属性失效
     * @see dimAmount
     */
    fun dimBehind(dimBehind: Boolean): AnyDialog {
        this.dimBehind = dimBehind
        return this
    }

    /**
     * 背景变暗，所有dialog公用同一个背景遮罩
     * 如果2个dialog的dimAmount不同，则以最后弹出的为准
     * 即，如果先弹出一个dimAmount=1的弹窗，背景会变为黑色
     * 再弹出一个dimAmount=0的弹窗，背景会变为透明，可以看到下面的Activity
     * 这时应该设置这个弹窗的dimBehind=false以达到透明背景效果
     * @see dimBehind
     */
    fun dimAmount(dimAmount: Float): AnyDialog {
        this.dimAmount = dimAmount
        return this
    }

    fun cancelable(cancelable: Boolean): AnyDialog {
        setCancelable(cancelable)
        setCanceledOnTouchOutside(cancelable)
        return this
    }

    fun cancelableOnClickKeyBack(cancelable: Boolean): AnyDialog {
        setCancelable(cancelable)
        return this
    }

    fun cancelableOnTouchOutside(cancelable: Boolean): AnyDialog {
        setCanceledOnTouchOutside(cancelable)
        return this
    }

    fun click(vararg ids: Int, listener: AnyDialog.(view: View) -> Unit): AnyDialog {
        viewHolder.addClickListener(ids, listener)
        return this
    }

    fun clickDismiss(vararg ids: Int, listener: (AnyDialog.(view: View) -> Unit)? = null): AnyDialog {
        viewHolder.addClickListener(ids) {
            dismiss()
            listener?.invoke(this, it)
        }
        return this
    }

    fun bindData(dataBind: AnyDialog.() -> Unit): AnyDialog {
        this.dataBind = dataBind
        return this
    }

    fun onShow(listener: AnyDialog.() -> Unit): AnyDialog {
        setOnShowListener {
            listener.invoke(this)
        }
        return this
    }

    fun onDismiss(listener: AnyDialog.() -> Unit): AnyDialog {
        setOnDismissListener {
            listener.invoke(this)
        }
        return this
    }

    fun onKey(listener: AnyDialog.(code: Int, event: KeyEvent) -> Boolean): AnyDialog {
        setOnKeyListener { _, code, event ->
            listener.invoke(this, code, event)
        }
        return this
    }

    fun <V : View> find(@IdRes id: Int): V? {
        return viewHolder.find(id)
    }

    enum class Style(
            val animRes: Int,
            val gravity: Int,
            val width: Int?,
            val height: Int?,
            val dragDirection: DragDirection?
    ) {
        CENTER(R.style.DialogAnimDef, Gravity.CENTER,
                null, null,
                null),
        BOTTOM(R.style.DialogAnimBottom, Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL,
                WindowManager.LayoutParams.MATCH_PARENT, null,
                DragDirection.BOTTOM),
        TOP(R.style.DialogAnimTop, Gravity.TOP or Gravity.CENTER_HORIZONTAL,
                WindowManager.LayoutParams.MATCH_PARENT, null,
                DragDirection.TOP),
        LEFT(R.style.DialogAnimLeft, Gravity.LEFT or Gravity.CENTER_VERTICAL,
                null, WindowManager.LayoutParams.MATCH_PARENT,
                DragDirection.LEFT),
        RIGHT(R.style.DialogAnimRight, Gravity.RIGHT or Gravity.CENTER_VERTICAL,
                null, WindowManager.LayoutParams.MATCH_PARENT,
                DragDirection.RIGHT)
    }

    enum class DragDirection {
        LEFT, TOP, RIGHT, BOTTOM
    }
}